src/native/custom_key_op_handler.c

```diff
@@ -0,0 +1,170 @@
+    }
+    if (jni_operation) {
+        (*env)->DeleteLocalRef(env, jni_operation);
+    }
+    if (!success && !exception_occured) {
+        aws_tls_key_operation_complete_with_error(operation, AWS_ERROR_UNKNOWN);
+    }
+
+    /* Release the Java ENV */
+    aws_jni_release_thread_env(op_handler->jvm, env);
+}
+
+static void s_aws_custom_key_op_handler_destroy(struct aws_custom_key_op_handler *key_op_handler) {
+
+    struct aws_jni_custom_key_op_handler *op_handler = (struct aws_jni_custom_key_op_handler *)key_op_handler->impl;
+
+    /* Get the Java ENV */
+    JNIEnv *env = aws_jni_acquire_thread_env(op_handler->jvm);
+    if (env == NULL) {
+        /* JVM is likely shutting down. Do not crash but log error. */
+        AWS_LOGF_ERROR(
+            AWS_LS_COMMON_IO, "java_custom_key_op_handler=%p destroy: Could not get Java ENV!", (void *)op_handler);
+        return;
+    }
+
+    /* Release the global reference */
+    if (op_handler->jni_custom_key_op) {
+        (*env)->DeleteGlobalRef(env, op_handler->jni_custom_key_op);
+    }
+
+    /* Release the Java ENV */
+    aws_jni_release_thread_env(op_handler->jvm, env);
+
+    /* Release the Java struct */
+    aws_mem_release(op_handler->allocator, op_handler);
+}
+
+static struct aws_custom_key_op_handler_vtable s_aws_custom_key_op_handler_vtable = {
+    .on_key_operation = s_aws_custom_key_op_handler_perform_operation,
+};
+
+struct aws_jni_custom_key_op_handler *aws_custom_key_op_handler_java_new(
+    JNIEnv *env,
+    struct aws_allocator *allocator,
+    jobject jni_custom_key_op) {
+
+    struct aws_jni_custom_key_op_handler *java_custom_key_op_handler =
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_jni_custom_key_op_handler));
+    if (java_custom_key_op_handler == NULL) {
+        return NULL;
+    }
+
+    if ((*env)->GetJavaVM(env, &java_custom_key_op_handler->jvm) != 0) {
+        aws_jni_throw_runtime_exception(env, "failed to get JVM");
+        aws_mem_release(allocator, java_custom_key_op_handler);
+        return NULL;
+    }
+
+    aws_ref_count_init(
+        &java_custom_key_op_handler->key_handler.ref_count,
+        &java_custom_key_op_handler->key_handler,
+        (aws_simple_completion_callback *)s_aws_custom_key_op_handler_destroy);
+    java_custom_key_op_handler->key_handler.vtable = &s_aws_custom_key_op_handler_vtable;
+    java_custom_key_op_handler->key_handler.impl = (void *)java_custom_key_op_handler;
+    /* Make a global reference so the Java interface is kept alive */
+    java_custom_key_op_handler->jni_custom_key_op = (*env)->NewGlobalRef(env, jni_custom_key_op);
+    java_custom_key_op_handler->allocator = allocator;
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_COMMON_IO,
+        "java_custom_key_op_handler=%p: Initalizing Custom Key Operations",
+        (void *)java_custom_key_op_handler);
+    return java_custom_key_op_handler;
+}
+
+void aws_custom_key_op_handler_java_release(
+    struct aws_allocator *allocator,
```

Comment:
you do not need to pass an `allocator` to this function

just call aws_jni_get_allocator()